### PR TITLE
Correctly detect serializable types from metadata using reflection

### LINKED ISSUE #1510: False positives in CA2235
#### Analyzer package

Microsoft Code Analysis, version 2.6.0.6241303

#### Analyzer

CA2235

#### Repro steps

In a .NET 4.6.1 project, add this class:

    using System;

    [Serializable]
    class C
    {
        DateTime date;
    }

#### Expected behaviour

No CA2235 diagnostics.

#### Actual behaviour

A diagnostic is reported with message `Field date is a member of type C which is serializable but is of type System.DateTime which is not serializable`

Seems that #261 may be back.

### PR DESCRIPTION
**Pull request by mavasani**
https://github.com/dotnet/roslyn/issues/3898 prevents the existing attribute based detection code to work for large number of cases. This PR adds a workaround to use reflection to read this property. Once the Roslyn issue is fixed, this code must be removed.

Fixes #1510

### CODE REVIEW

**Comment by mavasani** [comment]
Filed https://github.com/dotnet/roslyn-analyzers/issues/1520 to revert this temporary workaround once dotnet/roslyn#3898 is fixed.

File: src/Microsoft.NetFramework.Analyzers/Core/SerializationRulesDiagnosticAnalyzer.cs [hunk 1]
```diff
@@ -264,34 +266,20 @@ private bool IsSerializable(ITypeSymbol type)
...
+                        if (type.GetAttributes().Any(a => a.AttributeClass.Equals(_serializableAttributeTypeSymbol)))
+                        {
+                            return true;
+                        }
+
+                        // Workaround for https://github.com/dotnet/roslyn/issues/3898
+                        var isSerializableMetadataProperty = type.GetType().GetRuntimeProperties().FirstOrDefault(p => p.Name.Equals("IsSerializable", StringComparison.Ordinal));
+                        if (isSerializableMetadataProperty != null &&
+                            isSerializableMetadataProperty.GetValue(type) is bool isSerializable &&
+                            isSerializable)
```
Comment (ivanbasov) [inline src/Microsoft.NetFramework.Analyzers/Core/SerializationRulesDiagnosticAnalyzer.cs]: nit: you can put this condition into the Where clause. Then, you do not need to extract the FirstOrDefault instance.

File: src/Microsoft.NetFramework.Analyzers/Core/SerializationRulesDiagnosticAnalyzer.cs [hunk 2]
```diff
@@ -264,34 +266,20 @@ private bool IsSerializable(ITypeSymbol type)
...
-            }
+                        if (type.GetAttributes().Any(a => a.AttributeClass.Equals(_serializableAttributeTypeSymbol)))
+                        {
+                            return true;
+                        }
+
+                        // Workaround for https://github.com/dotnet/roslyn/issues/3898
+                        var isSerializableMetadataProperty = type.GetType().GetRuntimeProperties().FirstOrDefault(p => p.Name.Equals("IsSerializable", StringComparison.Ordinal));
+                        if (isSerializableMetadataProperty != null &&
+                            isSerializableMetadataProperty.GetValue(type) is bool isSerializable &&
```
Comment (jinujoseph) [inline src/Microsoft.NetFramework.Analyzers/Core/SerializationRulesDiagnosticAnalyzer.cs]: IsSerializable check is intentional?

Comment (mavasani) [reply src/Microsoft.NetFramework.Analyzers/Core/SerializationRulesDiagnosticAnalyzer.cs]: Yes.

**Pull request merged**